File: custom_renewal/custom_renewal/report/customer_without_contacts/customer_without_contacts.py
# Copyright (c) 2025, lakshman and contributors
# For license information, please see license.txt

# import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data

import frappe
def execute(filters=None):
	columns = [
		{
            "label": "Select",
            "fieldname": "select_row",
            "fieldtype": "Check",
            "width": 60
        },
		{
			"label": "Customer", 
			"fieldname": "customer",
			"fieldtype": "Link", 
			"options": "Customer", 
			"width": 300
		},

		{
			"label": "Sales Person", 
			"fieldname": "sales_person",
			"fieldtype": "Data", 
			"width": 200
		},
	]

	conditions, values = get_conditions(filters)

	data = frappe.db.sql("""
		SELECT tc.name as customer, tst.sales_person
		FROM tabCustomer tc
		inner join `tabSales Team` tst on tst.parent = tc.name
		 {conditions}
		GROUP BY tc.name 
		ORDER BY tc.name
	""".format(conditions=conditions), values, as_dict=True)


	currency = filters.presentation_currency or frappe.get_cached_value(
		"Company", filters.company, "default_currency"
	)

	# Initialize select_row to 0 for all rows
	for row in data:
		row["select_row"] = 0
	#report_summary = get_report_summary(filters,columns, currency, data)
	# frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(report_summary)))
	#chart = get_chart_data(filters, columns, data)
	# frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(chart)))

	#return columns, data, None,chart, report_summary
	return columns, data

def get_conditions(filters):
	# frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(filters)))
	conditions = []
	values = {}

	if filters.get("party_name"):
		conditions.append("AND tc.name = %(party_name)s")
		values["party_name"] = filters.get("party_name")

	if filters.get("sales_person"):
		conditions.append("AND tst.sales_person = %(sales_person)s")
		values["sales_person"] = filters.get("sales_person")    

	return " ".join(conditions), values



def get_report_summary(filters,columns, currency, data):
	customer_count = 0
	
	# frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(new)))

	if data:
		for period in data:
			if period.customer :
				customer_count += 1		

	customer_label = ("Customer Count")
	
	return [
		{"value": str(customer_count),"indicator": "Green", "label": customer_label, "datatype": "Data"},
		
	]


def get_chart_data(filters,columns, data):
	customer_count = 0
	
	# frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(data)))

	if data:
		for period in data:
			if period.customer :
				customer_count += 1	

	
	return {
		"data": {
			"labels": "Customers Count",  # show max of 30 items in chart
			"datasets": [{"values": customer_count}],
		},
		"type": "pie",
		"colors":["#c80064"],
	}

File: custom_renewal/custom_renewal/report/customer_without_contacts/customer_without_contacts.js
// Copyright (c) 2025, lakshman and contributors
// For license information, please see license.txt
/* eslint-disable */

/*frappe.query_reports["Customer Without Contacts"] = {
	"filters": [

	]
};*/


frappe.query_reports["Customer Without Contacts"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "party_name",
			"label": __("Party"),
			"fieldtype": "MultiSelectList",
			"options": "Customer",
			get_data: function (txt) {
				return frappe.db.get_link_options('Customer', txt);
			}
		},
		{
			"fieldname": "sales_person",
			"label": __("Sales Person"),
			"fieldtype": "MultiSelectList",
			"options": "Sales Person",
			get_data: function (txt) {
				return frappe.db.get_link_options('Sales Person', txt);
			}
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "select_row") {
			const checked = data.select_row ? 'checked' : '';
			return `<input type="checkbox" class="report-checkbox" data-row='${data.customer}' ${checked}>`;
		}
		return value;
	},

	onload: function (report) {
		let assignBtn;

		function setCheckboxes(checked = true) {
			const dt = report.datatable;
			if (!dt || !dt.wrapper) {
				frappe.msgprint("No data available.");
				return;
			}

			// Update internal row data
			if (dt.datamanager?.data) {
				dt.datamanager.data.forEach(row => {
					row.select_row = checked ? 1 : 0;
				});
			}

			// Refresh datatable to reflect checkbox state
			dt.refresh();

			updateAssignButtonVisibility();
		}

		function updateAssignButtonVisibility() {
			const anyChecked = !!document.querySelector('.report-checkbox:checked');
			if (assignBtn) {
				assignBtn.toggle(anyChecked);
			}
		}

		report.page.add_inner_button(__('Select All'), () => {
			setCheckboxes(true);
		});

		report.page.add_inner_button(__('Clear Selection'), () => {
			setCheckboxes(false);
		});

		assignBtn = report.page.add_inner_button("Assign User", function () {
			let selected = [];
			$('.report-checkbox:checked').each(function () {
				selected.push($(this).attr("data-row"));
			});

			if (!selected.length) {
				frappe.msgprint("Please select at least one customer.");
				return;
			}

			frappe.prompt([
				{
					label: 'User Email',
					fieldname: 'user_email',
					fieldtype: 'Link',
					options: 'User',
					reqd: 1
				}
			], function (values) {
				frappe.call({
					method: 'custom_renewal.api.bulk_assign_customer_to_salesperson_by_user',
					args: {
						customers: selected,
						user_email: values.user_email
					},
					callback: function (r) {
						if (r.message === "ok") {
							frappe.msgprint("User assigned successfully.");
							report.refresh();
						} else {
							frappe.msgprint("Something went wrong: " + (r.message || "No response."));
						}
					}
				});
			}, 'Assign User', 'Assign');
		});

		assignBtn.hide(); // initially hidden

		// Sync checkbox UI with datamanager data
		$(document).on('change', '.report-checkbox', function () {
			const rowKey = $(this).attr("data-row");
			const isChecked = $(this).is(":checked");

			const dt = report.datatable;
			if (dt?.datamanager?.data) {
				let row = dt.datamanager.data.find(r => r.customer === rowKey);
				if (row) {
					row.select_row = isChecked ? 1 : 0;
				}
			}

			updateAssignButtonVisibility();
		});
	}
};
